### PR TITLE
Support parameters in translation keys

### PR DESCRIPTION
**Pull request by jlsjonas**
Closes https://github.com/laravel-enso/Localisation/issues/51

Example Usage: (js counterpart of [Laravel's implementation](https://laravel.com/docs/5.7/localization#replacing-parameters-in-translation-strings))

`__('Hello :name :Name :NAME', {name:this.$store.state.user.person.name})`  
Given "jonas" as name it would translate (by default) to `Hello jonas Jonas JONAS`

Please also see notes in https://github.com/laravel-enso/Localisation/issues/51#issuecomment-444149731 regarding `addMissingKeys`

### CODE REVIEW

File: src/resources/js/modules/enso/mixins/__.js [hunk 1]
```diff
@@ -5,19 +5,12 @@ const __ = store.getters['localisation/__'];
 
 Vue.mixin({
     methods: {
-        __(key) {
+        __(key, params) {
             if (!store.getters['localisation/isInitialised']) {
                 return key;
             }
 
-            const translation = __(key);
```
Comment (aocneanu) [inline src/resources/js/modules/enso/mixins/__.js]: Let's leave this logic in the helper

File: src/resources/js/store/localisation.js [hunk 2]
```diff
@@ -6,11 +6,32 @@ export const state = {
 
 export const getters = {
     isInitialised: state => Object.keys(state.i18n).length > 0,
-    __: (state, getters, rootState, rootGetters) => (key) => {
-        const lang = rootGetters['preferences/lang'];
-        return state.i18n[lang]
+    __: (state, getters, rootState, rootGetters) => (key, params) => {
+        const lang = rootGetters['preferences/lang']
+            || ((rootState.preferences && rootState.preferences.global)
+                ? rootState.preferences.global.lang : null);
```
Comment (aocneanu) [inline src/resources/js/store/localisation.js]: I'm not sure that I understand why this ternary is needed

File: src/resources/js/store/localisation.js [hunk 3]
```diff
@@ -6,11 +6,32 @@ export const state = {
...
+
+        let translation = state.i18n[lang]
             ? state.i18n[lang][key]
             : key;
+
+        if (typeof translation === 'undefined'
+            && rootState.localisation.keyCollector) {
+                store.dispatch('localisation/addMissingKey', key);
+        }
```
Comment (aocneanu) [inline src/resources/js/store/localisation.js]: why should we call a mutation / action from a getter ?

File: src/resources/js/store/localisation.js [hunk 4]
```diff
@@ -6,11 +6,32 @@ export const state = {
...
+            translation = translation.replace(/:(\w*)/g, function(e, key) {
+                let param = params[key.toLowerCase()];
+                if(key === key.toUpperCase()) // param is uppercased
+                    param = param.toUpperCase();
+                else if(key[0] === key[0].toUpperCase()) // first letter is uppercased
+                    param = param.charAt(0).toUpperCase() + param.slice(1);
+                return param;
+            });
+        }
+        return translation;
```
Comment (aocneanu) [inline src/resources/js/store/localisation.js]: Let's move all the `params` logic to the global helper

**Comment by aocneanu** [comment]
I reviewed the PR.

Thanks for doing this!

File: src/resources/js/store/localisation.js (hunk 2 above)
Comment (jlsjonas) [reply src/resources/js/store/localisation.js]: I had a case (during login if I remember correctly) that, for some reason, the getter wasn't correctly returning the language but the rootstate was (which prevented the correct language from showing after navigating to the framework using a non-english language) I'll check if this is still needed

File: src/resources/js/store/localisation.js (hunk 4 above)
Comment (jlsjonas) [reply src/resources/js/store/localisation.js]: The _global_ helper is not global, it can't be called from within the store (ex. for a Dialog); hence why it's placed here instead (actually globally reachable)

The helper (imho) should be just that, an easier access to the store's `__`; not containing extra logic unless it only made sense in that context

File: src/resources/js/store/localisation.js (hunk 3 above)
Comment (jlsjonas) [reply src/resources/js/store/localisation.js]: As the _global_ helper isn't available within the vue store context itself (we have a few translatable strings directly in there as they're being used for _global_ dialogs)

**Comment by jlsjonas** [comment]
You're welcome

I commented with the reasons behind doing what I did; waiting for you reply before actually making changes

File: src/resources/js/store/localisation.js (hunk 2 above)
Comment (jlsjonas) [reply src/resources/js/store/localisation.js]: confirming the _fix_ is still required for translations to work after pageload

File: src/resources/js/store/localisation.js (hunk 3 above)
Comment (aocneanu) [reply src/resources/js/store/localisation.js]: I wasn't clear enough. We won't call mutations nor actions from getters. We will use the global helpers as it is, and if it is needed outside a vue instance we can import it.

File: src/resources/js/store/localisation.js (hunk 4 above)
Comment (aocneanu) [reply src/resources/js/store/localisation.js]: see above

Comment (jlsjonas) [reply src/resources/js/store/localisation.js]: For architectural reasons I definitely understand you don't want to call mutations/actions from getters (although this use-case is one of the few where it makes sense) but I don't see why you would move the params logic to the helper instead; otherwise you can't use parameterized translations from within the store (VueX is not a Vue component, see [Mixin docs](https://vuejs.org/v2/guide/mixins.html))

File: src/resources/js/store/localisation.js (hunk 3 above)
Comment (jlsjonas) [reply src/resources/js/store/localisation.js]: I understand, however see below for the implications

File: src/resources/js/store/localisation.js (hunk 4 above)
Comment (aocneanu) [reply src/resources/js/store/localisation.js]: I posted in the related issue the solution that will keep the store cleaner and solve your problem.

Comment (aocneanu) [reply src/resources/js/store/localisation.js]: >VueX is not a Vue component

sharing revelations I guess? :)

Comment (jlsjonas) [reply src/resources/js/store/localisation.js]: > sharing revelations I guess? :)

No, clarifying why what you suggested wasn't the solution ;)

**Comment by jlsjonas** [comment]
Note that Codacy is still failing only due to function complexity which is required in this case.

File: src/resources/js/store/localisation.js (hunk 4 above)
Comment (jlsjonas) [reply src/resources/js/store/localisation.js]: This has implications for the page title btw:
- param names would just be shown instead of actual data
- (confirmed on laravel-enso.com) page titles show up as null or undefined when there's no translation

The _fix_ with these changes (importing the _external_ `__` from within localisation.js) feels like an anti-pattern, open to suggestions though

File: src/resources/js/modules/enso/__.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+            if(key === key.toUpperCase()) { // param is uppercased
+                param = param.toUpperCase();
+            } else if(key[0] === key[0].toUpperCase()) { // first letter is uppercased
+                param = param.charAt(0).toUpperCase() + param.slice(1);
+            }
+            return param;
+        });
+    }
+    return translation;
+}
```
Comment (aocneanu) [inline src/resources/js/modules/enso/__.js]: I did a small refactor that covers some edge cases, but please test it (I suppose you have everything set for that)

```vue
import store from '../../store';

export default (key, params = null) => {
    if (!store.getters['localisation/isInitialised']) {
        return key;
    }

    let translation = store.getters['localisation/__'](key);

    if (typeof translation === 'undefined') {
        translation = key;

        if (store.state.localisation.keyCollector) {
           store.dispatch('localisation/addMissingKey', key);
        }
    }

    return !!params && typeof params === 'object'
        ? translation.replace(/:(\w*)/g, (e, key) => {
            if (typeof params[key.toLowerCase()] === 'undefined') {
                return key;
            }

            const param = params[key.toLowerCase()];

            if(key === key.toUpperCase()) {
                return param.toUpperCase();
            }
            
            return key[0] === key[0].toUpperCase()
                ? param.charAt(0).toUpperCase() + param.slice(1)
                : param;
            }
        })
        : translation;
}
```

File: src/resources/js/store/localisation.js [hunk 6]
```diff
@@ -6,11 +6,14 @@ export const state = {
 
 export const getters = {
     isInitialised: state => Object.keys(state.i18n).length > 0,
-    __: (state, getters, rootState, rootGetters) => (key) => {
-        const lang = rootGetters['preferences/lang'];
+    __: (state, getters, rootState, rootGetters) => (key, params) => {
```
Comment (aocneanu) [inline src/resources/js/store/localisation.js]: do we need `params` here?

File: src/resources/js/store/localisation.js [hunk 7]
```diff
@@ -6,11 +6,14 @@ export const state = {
...
+        const lang = rootGetters['preferences/lang']
+          || ((rootState.preferences && rootState.preferences.global)
+            ? rootState.preferences.global.lang : null);
+
         return state.i18n[lang]
-            ? state.i18n[lang][key]
-            : key;
```
Comment (aocneanu) [inline src/resources/js/store/localisation.js]: four spaces

File: src/resources/js/store/localisation.js [hunk 8]
```diff
@@ -6,11 +6,14 @@ export const state = {
 
 export const getters = {
     isInitialised: state => Object.keys(state.i18n).length > 0,
-    __: (state, getters, rootState, rootGetters) => (key) => {
-        const lang = rootGetters['preferences/lang'];
+    __: (state, getters, rootState, rootGetters) => (key, params) => {
+        const lang = rootGetters['preferences/lang']
+          || ((rootState.preferences && rootState.preferences.global)
+            ? rootState.preferences.global.lang : null);
+
```
Comment (aocneanu) [inline src/resources/js/store/localisation.js]: let's identify the cause for the problem you encountered and fix it at that level.

At a first sight the three lines are equivalent with

` const lang = rootGetters['preferences/lang'] || null;`

Comment (jlsjonas) [reply src/resources/js/store/localisation.js]: As mentioned, the issue is with a fresh-load when localisations are set, the translations don't show up.

My investigation indicated that, for some reason, the getter returns null although the state is already filled (although they should both return the exact same in theory)

File: src/resources/js/modules/enso/__.js (hunk 5 above)
Comment (jlsjonas) [reply src/resources/js/modules/enso/__.js]: Had to fix a syntax error & add a `== null` check to translation (missing keys are returning a null object instead of undefined), committing changes

File: src/resources/js/store/localisation.js (hunk 8 above)
Comment (jlsjonas) [reply src/resources/js/store/localisation.js]: Note that `null` should be `undefined` to avoid complications in the framework (document.title becomes null, and likely other hidden things)

Comment (aocneanu) [reply src/resources/js/store/localisation.js]: I'll try in the following days to replicate and see if what we have above is the best fix

**Comment by aocneanu** [comment]
Thanks again for the idea, time and patience!

**Comment by aocneanu** [comment]
Btw, 59f2bd1 fixes the doc title in Auth.

**Comment by jlsjonas** [comment]
You're welcome, glad I could help!

**Comment by jlsjonas** [comment]
🤔 something seems broken, unless it's a server-side caching/deployment issue (I reproduced it in private mode).
check laravel-enso.com, no matter what language I set the system to, I'm always just receiving english back (after clearing cache locally)

VueX is correctly showing the language objects & `preferences/lang` is set to non-english
@aocneanu

**Comment by aocneanu** [comment]
fixed in a38969f
